File: cpp/21_amicable_numbers.cpp
#include <iostream>
#include <cmath>

int get_divisor_sum(int n){
    int divSum = 1;
    int s = int(std::sqrt(n));
    for (int i = 2; i <= s; i++){
        if (n % i == 0){
            divSum += i + (n / i);
        }
    }
    if (s * s == n) divSum -= s;
    return divSum;
}

int amicable_numbers(){
    int amSum = 0;
    for (int n = 2; n < 10'000; n++){
        int divSum = get_divisor_sum(n);
        // count amicable numbers  once
        // a != b so either a > b or b > a
        if (divSum > n && get_divisor_sum(divSum) == n){
            amSum += n + divSum;
        }
    }
    return amSum;
}

int main() {
    std::cout << amicable_numbers() << "\n";
}

File: c#/_21_AmicableNumbers.cs
using System;

public class _21_AmicableNumbers{
	public static int Euler21(){
		int sum = 0;
		for (int i = 2; i < 10_000; i++){
			int divSum = GetProperDivisorSum(i);
			if (divSum > i && GetProperDivisorSum(divSum) == i){
				sum += i + divSum;
			}
		}
		return sum;
	}
	
	public static int GetProperDivisorSum(int n){
		if (n == 1) return 0;
		int res = 1;
		for (int i = 2; i <= Math.Sqrt(n); i++){
			if (n % i == 0){
				if (n / i == i) res += i;
				else res += i + n / i;
			}
		}
		return res;
	}
	
	public static void Main(){
		Console.WriteLine(Euler21());
	}
}
